cmd/rancher-machine/setup_test.go

```diff
@@ -0,0 +1,184 @@
+	// TestMain.
+	logCommandOutput bool
+
+	// logCommandOutputOnFail indicates whether to log command stdout and stderr if a test fails. This should be set in
+	// TestMain.
+	logCommandOutputOnFail bool
+
+	// executablePath is the path to the rancher-machine binary to be executed in tests.
+	executablePath string
+)
+
+// TestMain tests up and runs all tests in this package. The tests can be invoked as follows:
+//
+//  1. Build the rancher-machine binary that you'd like to test.
+//
+//  2. Run the tests. Note that the -v flag should be set on the "go test" command to ensure that go test doesn't
+//     swallow all command output.
+//
+//     If you want to test the binary at path/to/binary, logging all command output from tests:
+//     $ go test -v ./... -args -log -exec-path path/to/binary
+//
+//     If you want to run the rancher-machine binary in the default location (./rancher-machine) with default logging
+//     (only log on failure):
+//     $ go test -v ./...
+func TestMain(m *testing.M) {
+	// Check if we should log command output.
+	//
+	// Testing with all command logging enabled (-v required so go test doesn't capture output):
+	//
+	// 		$ go test -v ./... -args -log -exec-path path/to/rancher-machine
+	//
+	// Testing with only command logging on test failure (-v required so go test doesn't capture output):
+	//
+	//		$ go test -v ./... -args -log-fail -exec-path path/to/rancher-machine
+	flag.BoolVar(
+		&logCommandOutput,
+		"log",
+		false,
+		"Log all command stdout and stderr in tests (default: false)",
+	)
+	flag.BoolVar(
+		&logCommandOutputOnFail,
+		"log-fail",
+		true,
+		"Log all command stdout and stderr on test failures (default: true)",
+	)
+	flag.StringVar(
+		&executablePath,
+		"exec-path",
+		"./rancher-machine",
+		"Path to the rancher-machine binary to test (default: ./rancher-machine)",
+	)
+	flag.Parse()
+
+	// Get the current working directory.
+	curDir, err := os.Getwd()
+	if err != nil {
+		fmt.Printf("error getting working directory: %v", err)
+		os.Exit(1)
+	}
+
+	// Make sure that the rancher-machine binary is available in the current directory as it is invoked in tests.
+	stat, err := os.Stat(executablePath)
+	if err != nil || stat.IsDir() || !stat.Mode().IsRegular() {
+		fmt.Printf(
+			"invalid test executable %s: %v\n%s",
+			executablePath,
+			err,
+			"Use flag -exec-path to specify the path to the rancher-machine binary you want to test.",
+		)
+		os.Exit(1)
+	}
+
+	// If a test data directory already exists, clear it.
+	testDataPath = filepath.Join(curDir, "test-data")
+	if err = os.RemoveAll(testDataPath); err != nil {
+		fmt.Printf("error removing direcotry %s: %v", testDataPath, err)
+		os.Exit(1)
+	}
+
+	// Create a directory for temporary test data.
+	if err = os.MkdirAll(testDataPath, 0750); err != nil {
+		fmt.Printf("error creating direcotry %s: %v", testDataPath, err)
+		os.Exit(1)
+	}
+
+	// Run the test suite.
+	os.Exit(m.Run())
+}
```

Comment:
This is no blocker nor anything. For me this behavior is a bit strange. 

This code is creating a "temporary folder" for storing test data but it is not removing it.
That said the folder is being cleaned up on a second test execution. 

IMO: 
If the content of the folder is irrelevant we should delete it after the test execution. 
If the content of the folder is relevant we should never delete it and use sub-folders to enable keeping the record of more than one test.

Reply:
Good point. The reason I opted to not automatically delete the temp folder was so it can be used to debug. If your test fails, information about the state of the machines you create in your tests might be stored in that folder, so looking at the files in there could help you understand why the test is failing. I chose to simply overwrite that folder on subsequent test runs to avoid a situation where you have an ever-growing folder full of test data on your hard drive that never gets deleted.